promql/parse.go

```diff
@@ -461,26 +461,31 @@ func (p *parser) expr() Expr {
 			returnBool = true
 		}
 
-		// Parse ON clause.
-		if p.peek().typ == itemOn {
+		// Parse ON/IGNORING clause.
+		if p.peek().typ == itemOn || p.peek().typ == itemIgnoring {
+			if p.peek().typ == itemIgnoring {
+				vecMatching.Ignoring = true
+			}
 			p.next()
 			vecMatching.On = p.labels()
 
 			// Parse grouping.
-			if t := p.peek().typ; t == itemGroupLeft {
-				p.next()
-				vecMatching.Card = CardManyToOne
-				vecMatching.Include = p.labels()
-			} else if t == itemGroupRight {
+			if t := p.peek().typ; t == itemGroupLeft || t == itemGroupRight {
 				p.next()
-				vecMatching.Card = CardOneToMany
-				vecMatching.Include = p.labels()
+				if t == itemGroupLeft {
+					vecMatching.Card = CardManyToOne
+				} else {
+					vecMatching.Card = CardOneToMany
+				}
+				if p.peek().typ == itemLeftParen {
+					vecMatching.Include = p.labels()
+				}
 			}
 		}
 
 		for _, ln := range vecMatching.On {
 			for _, ln2 := range vecMatching.Include {
-				if ln == ln2 {
+				if ln == ln2 && !vecMatching.Ignoring {
 					p.errorf("label %q must not occur in ON and INCLUDE clause at once", ln)
```

Comment:
Unrelated, but the INCLUDE might be confusing to the users, should be GROUP_LEFT/GROUP_RIGHT clause.